src/providers/ipa/ipa_subdomains_ext_groups.c

```diff
@@ -812,6 +812,83 @@ errno_t ipa_get_trusted_memberships_recv(struct tevent_req *req, int *dp_error_o
     return EOK;
 }
 
+static char *get_rdn_name_from_dn_string(TALLOC_CTX *mem_ctx,
+                                         struct ldb_context *ldb,
+                                         const char *dn_str)
+{
+    struct ldb_dn *dn;
+    const char *tmp_str;
+    char *out = NULL;
+
+    dn = ldb_dn_new(mem_ctx, ldb, dn_str);
+    if (dn == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "ldb_dn_new failed.\n");
+        return NULL;
+    }
+
+    tmp_str = ldb_dn_get_rdn_name(dn);
+    if (tmp_str == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to get RDN name.\n");
+        goto done;
+    }
+
+    out = talloc_strdup(mem_ctx, tmp_str);
+    if (out == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to copy attribute name.\n");
+    }
+
+done:
+    talloc_free(dn);
+    return out;
+}
+
+static errno_t filter_groups_by_attribute_name(TALLOC_CTX *mem_ctx,
+                                               struct ldb_context *ldb,
+                                               char **groups,
+                                               const char *allowed_attr_name,
+                                               char ***filtered_groups)
+{
+    size_t c;
+    size_t d = 0;
+    char **tmp_list;
+    char *tmp_str;
+
+    if (groups == NULL) {
+        *filtered_groups = NULL;
+        return EOK;
+    }
+
+    for (c = 0; groups[c] != NULL; c++);
+
+    tmp_list = talloc_array(mem_ctx, char *, c + 1);
+    if (tmp_list == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "Failed to allocate memory for output list.\n");
+        return ENOMEM;
+    }
+
+    for (c = 0; groups[c] != NULL; c++) {
+        tmp_str = get_rdn_name_from_dn_string(tmp_list, ldb, groups[c]);
```

Comment:
It would be way more lightweight to make helper function to (schematically)
```
s = strchr(groups[c], '=');
strncasecmp(allowed_attr_name, groups[c], (s-groups[c]))
```
and return bool to avoid tons of malloc/free.

Reply:
If this is the best place to implement filtering, then maybe at least `talloc_steal()` could be used instead of `talloc_strdup()` for group DNs? IIUC, original groups list isn't used anyway (though hangs on the state context).